File: library/src/main/java/org/incendo/interfaces/utilities/CollapsablePaneMap.java
package org.incendo.interfaces.utilities;

import org.incendo.interfaces.pane.CompletedPane;

import java.util.Collections;
import java.util.Map;
import java.util.TreeMap;
import java.util.concurrent.ConcurrentSkipListMap;

public final class CollapsablePaneMap extends ConcurrentSkipListMap<Integer, CompletedPane> {

    private final int rows;
    private CompletedPane cachedPane = null;

    private CollapsablePaneMap(final int rows, final Map<Integer, CompletedPane> internal) {
        super(internal);
        this.rows = rows;
    }

    public static CollapsablePaneMap create(final int rows) {
        return new CollapsablePaneMap(rows, new TreeMap<>(Collections.reverseOrder()));
    }

    @Override
    public CompletedPane put(final Integer key, final CompletedPane value) {
        this.cachedPane = null;
        return super.put(key, value);
    }

    public CompletedPane collapse() {
        if (this.cachedPane != null) {
            return this.cachedPane;
        }

        CompletedPane pane = CompletedPane.filled(this.rows);

        for (CompletedPane layer : this.values()) {
            pane.putAll(layer);
        }

        this.cachedPane = pane;
        return pane;
    }
}

File: next/src/main/java/org/incendo/interfaces/next/utilities/CollapsablePaneMap.kt
package org.incendo.interfaces.next.utilities

import org.incendo.interfaces.next.pane.CompletedPane
import org.incendo.interfaces.next.pane.Pane
import org.incendo.interfaces.next.pane.convertToEmptyCompletedPaneAndFill

internal class CollapsablePaneMap private constructor(
    private val rows: Int,
    // pass in a pane from the view so that we can persist
    // ordering, used in the listeners.
    private val basePane: Pane,
    // privately pass in a map here so that we can use
    // super methods when overriding methods in the delegate.
    private val internal: MutableMap<Int, CompletedPane>
) : MutableMap<Int, CompletedPane> by internal {

    internal companion object {
        internal fun create(rows: Int, basePane: Pane) = CollapsablePaneMap(
            rows,
            basePane,
            sortedMapOf(Comparator.reverseOrder())
        )
    }

    private var cachedPane: CompletedPane? = null

    override fun put(key: Int, value: CompletedPane): CompletedPane? {
        cachedPane = null
        return internal.put(key, value)
    }

    internal fun collapse(): CompletedPane {
        cachedPane?.let { pane ->
            return pane
        }

        val pane = basePane.convertToEmptyCompletedPaneAndFill(rows)

        values.forEach { layer ->
            layer.forEach { row, column, value ->
                pane[row, column] = value
            }
        }

        return pane
    }
}
